### PR TITLE
AirbyteLib: add null cache and null writer

### PR DESCRIPTION
**Pull request by aaronsteers**
Note:

This should merge after:

- #34588

This is born out of a desire to get a better understanding of performance bottlenecks. This PR add a "NullCache" which does nothing at all with incoming records. 

The goal with this is to be able to get performance benchmarks on a source connector, with little or no slowdown from AirbyteLib, the file writer, and/or the SQL cache.

### CODE REVIEW

File: airbyte-lib/airbyte_lib/_file_writers/null.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
+# Copyright (c) 2023 Airbyte, Inc., all rights reserved.
+
+"""A Parquet cache implementation."""
```
Comment (aaronsteers) [inline airbyte-lib/airbyte_lib/_file_writers/null.py:3]:
```suggestion
"""A No-Op (Null) cache implementation.

This is designed to test the raw throughput of a source, without any slowdown from writing files.
"""
```

File: airbyte-lib/airbyte_lib/_file_writers/null.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+
+class NullWriterConfig(FileWriterConfigBase):
+    """Configuration for the Snowflake cache."""
+
+    # Inherits `cache_dir` from base class
+
+
+class NullWriter(FileWriterBase):
+    """A Parquet cache implementation."""
```
Comment (aaronsteers) [inline airbyte-lib/airbyte_lib/_file_writers/null.py:30]:
```suggestion
class NullWriterConfig(FileWriterConfigBase):
    """Configuration for the Null (no-op) file writer."""

    # Inherits `cache_dir` from base class


class NullWriter(FileWriterBase):
    """A Null (no-op) file writer implementation."""
```

File: airbyte-lib/airbyte_lib/caches/null.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+    def get_database_name(self) -> str:
+        return "dummy"
+
+
+class NullCache(SQLCacheBase):
+    """A DuckDB implementation of the cache.
+
+    Parquet is used for local file storage before bulk loading.
+    Unlike the Snowflake implementation, we can't use the COPY command to load data
+    so we insert as values instead.
```
Comment (alafanechere) [inline airbyte-lib/airbyte_lib/caches/null.py:53]: I think this docstring is not up to date

File: airbyte-lib/airbyte_lib/caches/null.py [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+
+    @overrides
+    def get_telemetry_info(self) -> CacheTelemetryInfo:
+        return CacheTelemetryInfo("null")
+
+    @overrides
+    def _execute_sql(self, sql: str) -> None:
+        """Execute SQL."""
+        _ = sql
+        # Do nothing
```
Comment (alafanechere) [inline airbyte-lib/airbyte_lib/caches/null.py:68]: I'd appreciate an explicit "return None"

File: airbyte-lib/airbyte_lib/caches/null.py [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+
+    @overrides
+    def _ensure_compatible_table_schema(
+        self, stream_name: str, table_name: str, *, raise_on_error: bool = False
+    ) -> bool:
+        _ = stream_name, table_name, raise_on_error
+        pass
+
+    @overrides
+    def _write_files_to_new_table(self, files: list[Path], stream_name: str, batch_id: str) -> str:
```
Comment (alafanechere) [inline airbyte-lib/airbyte_lib/caches/null.py:89]: please change the type hint if none is returned here

**Review by alafanechere — CHANGES_REQUESTED** [review]

**Comment by flash1293** [comment]
Based on the stated goal of this PR - the best you should be able to do is to just use `list(source.get_records())`. It's basically 

> with little or no slowdown from AirbyteLib, the file writer, and/or the SQL cache.

**Comment by natikgadzhi** [comment]
I'm closing because I assume you already made this change in pyairbyte repo. If not, apologies! Seems like a huge lift to rebase and rename anyway.

**Pull request closed without merging**